// testing-library-selector/build/main/lib/selector.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.byTitle = exports.byPlaceholderText = exports.byTestId = exports.byDisplayValue = exports.byAltText = exports.byLabelText = exports.byBoundAttribute = exports.byText = exports.byRole = void 0;
const DomQueries = __importStar(require("@testing-library/dom"));
const dom_1 = require("@testing-library/dom");
function makeRTLSelector(selection) {
    // too lazy to list out all combinations... macros would come in handy here!
    const exec = (type, waitForOptions, container) => {
        const methodName = `${type}${selection.query === 'ByBoundAttribute'
            ? `By${selection.attribute}`
            : selection.query}`;
        const args = [
            selection.matcher,
            selection.options,
            ...(waitForOptions ? [waitForOptions] : []),
        ];
        if (container) {
            // @ts-ignore
            return DomQueries[methodName](container, ...args);
        }
        // @ts-ignore
        return dom_1.screen[methodName](...args);
    };
    return {
        get: (container) => exec('get', undefined, container),
        getAll: (container) => exec('getAll', undefined, container),
        find: (container, waitForOptions) => exec('find', waitForOptions, container),
        findAll: (container, waitForOptions) => exec('findAll', waitForOptions, container),
        query: (container) => exec('query', undefined, container),
        queryAll: (container) => exec('queryAll', undefined, container),
    };
}
function byRole(role, options) {
    return makeRTLSelector({ query: 'ByRole', matcher: role, options });
}
exports.byRole = byRole;
function byText(id, options) {
    return makeRTLSelector({ query: 'ByText', matcher: id, options });
}
exports.byText = byText;
function byBoundAttribute(attribute, id, options) {
    return makeRTLSelector({
        query: 'ByBoundAttribute',
        attribute,
        matcher: id,
        options,
    });
}
exports.byBoundAttribute = byBoundAttribute;
function byLabelText(id, options) {
    return makeRTLSelector({ query: 'ByLabelText', matcher: id, options });
}
exports.byLabelText = byLabelText;
function byAltText(id, options) {
    return byBoundAttribute('AltText', id, options);
}
exports.byAltText = byAltText;
function byDisplayValue(id, options) {
    return byBoundAttribute('DisplayValue', id, options);
}
exports.byDisplayValue = byDisplayValue;
function byTestId(id, options) {
    return byBoundAttribute('TestId', id, options);
}
exports.byTestId = byTestId;
function byPlaceholderText(id, options) {
    return byBoundAttribute('PlaceholderText', id, options);
}
exports.byPlaceholderText = byPlaceholderText;
function byTitle(id, options) {
    return byBoundAttribute('Title', id, options);
}
exports.byTitle = byTitle;
