// testing-library-selector/build/module/lib/selector.js
import * as DomQueries from '@testing-library/dom';
import { screen, } from '@testing-library/dom';
function makeRTLSelector(selection) {
    // too lazy to list out all combinations... macros would come in handy here!
    const exec = (type, waitForOptions, container) => {
        const methodName = `${type}${selection.query === 'ByBoundAttribute'
            ? `By${selection.attribute}`
            : selection.query}`;
        const args = [
            selection.matcher,
            selection.options,
            ...(waitForOptions ? [waitForOptions] : []),
        ];
        if (container) {
            // @ts-ignore
            return DomQueries[methodName](container, ...args);
        }
        // @ts-ignore
        return screen[methodName](...args);
    };
    return {
        get: (container) => exec('get', undefined, container),
        getAll: (container) => exec('getAll', undefined, container),
        find: (container, waitForOptions) => exec('find', waitForOptions, container),
        findAll: (container, waitForOptions) => exec('findAll', waitForOptions, container),
        query: (container) => exec('query', undefined, container),
        queryAll: (container) => exec('queryAll', undefined, container),
    };
}
export function byRole(role, options) {
    return makeRTLSelector({ query: 'ByRole', matcher: role, options });
}
export function byText(id, options) {
    return makeRTLSelector({ query: 'ByText', matcher: id, options });
}
export function byBoundAttribute(attribute, id, options) {
    return makeRTLSelector({
        query: 'ByBoundAttribute',
        attribute,
        matcher: id,
        options,
    });
}
export function byLabelText(id, options) {
    return makeRTLSelector({ query: 'ByLabelText', matcher: id, options });
}
export function byAltText(id, options) {
    return byBoundAttribute('AltText', id, options);
}
export function byDisplayValue(id, options) {
    return byBoundAttribute('DisplayValue', id, options);
}
export function byTestId(id, options) {
    return byBoundAttribute('TestId', id, options);
}
export function byPlaceholderText(id, options) {
    return byBoundAttribute('PlaceholderText', id, options);
}
export function byTitle(id, options) {
    return byBoundAttribute('Title', id, options);
}
